Allow l2tpd_t access to netlink and sysfs

tomparkin opened the pull request:
The go-l2tp kl2tpd daemon used by NetworkManager-l2tp uses netlink_generic_socket and sysfs.

This change addresses the following AVC denials:

type=AVC msg=audit(1721045130.932:277): avc:  denied  { read } for  pid=3560 comm="kl2tpd" name="hpage_pmd_size" dev="sysfs" ino=1261 scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:object_r:sysfs_t:s0 tclass=file permissive=1
type=AVC msg=audit(1721045130.932:278): avc:  denied  { open } for  pid=3560 comm="kl2tpd" path="/sys/kernel/mm/transparent_hugepage/hpage_pmd_size" dev="sysfs" ino=1261 scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:object_r:sysfs_t:s0 tclass=file permissive=1
type=AVC msg=audit(1721045130.942:279): avc:  denied  { create } for  pid=3560 comm="kl2tpd" scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:system_r:l2tpd_t:s0 tclass=netlink_generic_socket permissive=1
type=AVC msg=audit(1721045130.942:280): avc:  denied  { getopt } for  pid=3560 comm="kl2tpd" scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:system_r:l2tpd_t:s0 tclass=netlink_generic_socket permissive=1
type=AVC msg=audit(1721045130.942:281): avc:  denied  { bind } for  pid=3560 comm="kl2tpd" scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:system_r:l2tpd_t:s0 tclass=netlink_generic_socket permissive=1
type=AVC msg=audit(1721045130.942:282): avc:  denied  { getattr } for  pid=3560 comm="kl2tpd" scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:system_r:l2tpd_t:s0 tclass=netlink_generic_socket permissive=1

Resolves: https://github.com/fedora-selinux/selinux-policy/issues/2259

zpytela commented on policy/modules/contrib/l2tp.te:
```diff
@@ -30,12 +30,14 @@ allow l2tpd_t self:capability net_admin;
 allow l2tpd_t self:unix_dgram_socket sendto;
 allow l2tpd_t self:unix_stream_socket { accept listen };
 allow l2tpd_t self:pppox_socket create_socket_perms;
+allow l2tpd_t sysfs_t:file { open read };
```
This line is a subset of dev_read_sysfs() which actually is the correct way to access types from other modules.

tomparkin replied:
Sorry, my mistake.

I thought the same from reading the macro definition for _dev_read_sysfs()_, but during development I found I needed the extra "allow" line otherwise I was still hitting some access denials.

However I think I must have gotten my packages mixed up in the test environment as I've just re-tested from scratch without the "allow" and and everything is working.  I'll remove this extra line.

zpytela reviewed (changes requested):
If you know anything about the code changes, please add them to the commit message or add a link.

tomparkin commented:
Thanks @zpytela for reviewing.

Regarding commit comments, what sort of extra detail would be appropriate?

I tried to mirror the commit comment style in the git log, and the github issue has extra context.  Should I add information about kl2tpd requirements, etc?

dkosovic commented:
> Regarding commit comments, what sort of extra detail would be appropriate?
> 
> I tried to mirror the commit comment style in the git log, and the github issue has extra context. Should I add information about kl2tpd requirements, etc?

I wonder if @zpytela is referring to adding a higher-level background to the commit message as to why this pull request is required, e.g.:

The go-l2tp kl2tpd daemon is in the `golang-github-katalix-l2tp` RPM which was first introduced with Fedora 41:
- https://src.fedoraproject.org/rpms/golang-github-katalix-l2tp

NetworkManager-l2tp version 1.20.0 and later use `kl2tpd` as the preferred L2TP daemon if installed,
otherwise will fallback to using `xl2tpd` :
- https://github.com/nm-l2tp/NetworkManager-l2tp/releases/tag/1.20.0

Newer NetworkManager-l2tp RPMs will have the `xl2tpd` requires dependency changed to `(go-l2tp or xl2tpd)`
to ensure `kl2tpd` gets installed by default, but still allow `xl2tpd`.

maksimsamt commented:
@tomparkin , @zpytela ,

With Fedora 43 this issue becomes critical:
https://bugzilla.redhat.com/show_bug.cgi?id=2406859

Is it possible to adjust this pull request according to this semodule config?
```
module kl2tpd-custom 1.0;
require {
  type l2tpd_t;
  type sysfs_t;
  class file { open read };
  class netlink_generic_socket { bind create getattr getopt read write };
}
#============= l2tpd_t ==============
allow l2tpd_t self:netlink_generic_socket { bind create getattr getopt read write };
allow l2tpd_t sysfs_t:file { open read };
```

And when possible push the changes for the next Fedora update?

zpytela commented:
@tomparkin Will you be able to address my comments, i.e. remove direct access to sysfs_t and add some comment to the commit?

There seems to be new demand for these policy changes.

tomparkin commented:
@zpytela will do -- in terms of the commit comment I'm still not entirely sure what you'd like to see there but I can add some extra context about the usecase, etc, if that is appropriate.

zpytela commented:
Not much, just in general justification for the changes, or anything you know.

tomparkin committed:
Allow l2tpd_t access to netlink and sysfs

Layer 2 Tunnelling Protocol (L2TP) is a networking protocol supported
by the Linux kernel and used in L2TP/IPSec VPN implementations.

Although the kernel provides a dataplane for L2TP connections, the
control plane is implemented in userspace, with communications between
the kernel and userspace typically handled via. netlink.

One such L2TP userspace daemon is kl2tpd, which is included in Fedora
as a part of the go-l2tp[1] package, and is used by the NetworkManager-l2tp
plugin.

In order to run under selinux, kl2tpd needs access to netlink and
sysfs[2].

This change addresses the following AVC denials:

type=AVC msg=audit(1721045130.932:277): avc:  denied  { read } for  pid=3560 comm="kl2tpd" name="hpage_pmd_size" dev="sysfs" ino=1261 scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:object_r:sysfs_t:s0 tclass=file permissive=1
type=AVC msg=audit(1721045130.932:278): avc:  denied  { open } for  pid=3560 comm="kl2tpd" path="/sys/kernel/mm/transparent_hugepage/hpage_pmd_size" dev="sysfs" ino=1261 scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:object_r:sysfs_t:s0 tclass=file permissive=1
type=AVC msg=audit(1721045130.942:279): avc:  denied  { create } for  pid=3560 comm="kl2tpd" scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:system_r:l2tpd_t:s0 tclass=netlink_generic_socket permissive=1
type=AVC msg=audit(1721045130.942:280): avc:  denied  { getopt } for  pid=3560 comm="kl2tpd" scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:system_r:l2tpd_t:s0 tclass=netlink_generic_socket permissive=1
type=AVC msg=audit(1721045130.942:281): avc:  denied  { bind } for  pid=3560 comm="kl2tpd" scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:system_r:l2tpd_t:s0 tclass=netlink_generic_socket permissive=1
type=AVC msg=audit(1721045130.942:282): avc:  denied  { getattr } for  pid=3560 comm="kl2tpd" scontext=system_u:system_r:l2tpd_t:s0 tcontext=system_u:system_r:l2tpd_t:s0 tclass=netlink_generic_socket permissive=1

[1] https://src.fedoraproject.org/rpms/go-l2tp
[2] https://bugzilla.redhat.com/show_bug.cgi?id=2407022

Resolves: #2259

tomparkin commented:
@zpytela apologies for the delay, I have been busy with other commitments.

However, I have now updated this PR as requested, and rebased on current `rawhide`  -- I have force-pushed the update to keep it as a part of this discussion thread however if you'd prefer me to create a new PR I am happy to do so.

I note CI is currently failing, which I will look into now.

tomparkin commented:
> I note CI is currently failing, which I will look into now.

Hmm, I get the same failure on F43 building current tip of the `rawhide` branch without my changes, which leads me to suspect there's an issue elsewhere.

dkosovic commented:
There was another kl2tpd merge request https://github.com/fedora-selinux/selinux-policy/pull/3001 for the following Fedora Bugzilla report:
- https://bugzilla.redhat.com/show_bug.cgi?id=2422239

So, the latest instance of this merge request now has two instances of the following line:
```te
allow l2tpd_t self:netlink_generic_socket create_socket_perms;
```

Not sure if the duplicated line is related to the CI issue.

zpytela commented:
Duplicate lines do not matter, but I reverted the other commit.
Failing CI seems to be due to different environment, I haven't managed to find what is the root cause yet.
```Creating policy.xml
echo '<?xml version="1.0" encoding="ISO-8859-1" standalone="no"?>' > doc/policy.xml
echo '<!DOCTYPE policy SYSTEM "policy.dtd">' >> doc/policy.xml
echo '<policy>' >> doc/policy.xml
for i in admin apps contrib kernel roles services system; do echo "<layer name=\"$i\">" >> doc/policy.xml; cat tmp/$i.xml >> doc/policy.xml; echo "</layer>" >> doc/policy.xml; done
cat doc/global_tunables.xml doc/global_booleans.xml >> doc/policy.xml
echo '</policy>' >> doc/policy.xml
if test -x /usr/bin/xmllint && test -f doc/policy.dtd; then \
        /usr/bin/xmllint --noout --path doc/ --dtdvalid doc/policy.dtd doc/policy.xml ;\
fi
doc/policy.xml:8428: parser error : xmlParseEntityRef: no name
(public_content_t & public_content_rw_t)
                   ^
make: *** [Makefile:429: doc/policy.xml] Error 1
```

zpytela commented:
Merging, thank you.
